### PR TITLE
ref(discover) Consolidate snuba alias resolution

### PR DESCRIPTION
**Pull request by markstory**
Currently we resolve public facing snuba column aliases in a few different places:

* When parsing a query string expression
* When building queries with `dataset_query()`
* When using evenstore (sometimes)

These places are all at different layers in the application and to work around this we've added boolean flags to various functions to indicate that conditions and filter keys have already had their aliases resolved.

These changes remove all the alias resolution except for those done in `dataset_query()`. This function has all the context about a query and can resolve the field aliases in a dataset aware manner.

Refs SEN-1136

### CODE REVIEW

**Review by dashed — COMMENTED** [review]
Tests look fine to me. I'll defer to @lynnagara for a more thorough review.

File: src/sentry/utils/snuba.py [hunk 1]
```diff
@@ -1023,6 +1010,11 @@ def constrain_condition_to_dataset(cond, dataset):
     """
     index = get_function_index(cond)
     if index is not None:
+        # IN conditions are detected as a function but aren't really.
+        if cond[index] == "IN":
```
Comment (lynnagara) [inline src/sentry/utils/snuba.py]: Is this related to alias resolution?

**Review by lynnagara — COMMENTED** [review]
I previously understood dataset_query to be temporary code that figures out which dataset to use prior to this responsibility moving into Snuba, rather than something we want to now use everywhere we need to do alias resolution. This PR means that we will have to use dataset_query everywhere just for alias resolution even when we are not actually attempting to select a dataset. Would an alternative approach be to extract the alias resolution bits into something more reusable?

File: src/sentry/utils/snuba.py (hunk 1 above)
Comment (markstory) [reply src/sentry/utils/snuba.py]: Yes. `environment` conditions used to be pre-parsed and aliased by the event_search parser and would not have their columns re-aliased because of `skip_conditions`. Now, because all conditions are re-aliased the `IN` string is followed by a list which quacks like a function `get_function_index`.

**Comment by lynnagara** [comment]
We can still have a single path for translating fieldnames, but I don't think we should couple it in the same function with the dataset selection logic if possible.

For example, Eventstore never operates off the aliased fields, https://github.com/getsentry/sentry/blob/master/src/sentry/eventstore/base.py#L9-L67. 
If we add the transactions columns to this list and use it with detect_dataset (which I'd planned to do soon), the decision would be wrong because it won't handle unaliased columns.

File: src/sentry/utils/snuba.py (hunk 1 above)
Comment (dashed) [reply src/sentry/utils/snuba.py]: For posterity, I've added some tests to surface this issue in https://github.com/getsentry/sentry/pull/15214

**Comment by lynnagara** [comment]
> If we don't choose the dataset based on the public field names how can we resolve the aliases to the dataset specific name? Until the merged dataset is available I think we need to resolve the dataset and then resolve field aliases.
Is there a reason we can't resolve the dataset based on the internal names now? Since dataset selection is moving  into Snuba, but maintaining the list of external aliases will stay in Sentry, it seems like dataset choice would be based on the internal names at some point at least.

I agree that a Query abstraction that is dataset aware and can resolve the aliases is a very good idea.

**Comment by fpacifici** [comment]
About external/internal names, when we will move to a snuba merged dataset (which I think is something we could start building in a matter of days), this dataset will have a set of columns names that will look like more the external ones than the internal one. 
In that environment dataset selection becomes table selection. Since different tables have columns that mean the same thing with different name and type (as today happen between datasets), this resolution needs to happen starting from a column alias that is closer to the external one.
So in practice sentry is not going to need to do any table specific resolution. Now alias resolution existed before the transaction dataset for other reasons, and that can still exist if needed.

The alternative to this approach would be to have a merged dataset that will have the union of the columns of events and transaction, thus making the existence of a merged dataset almost pointless.

About the Query abstraction:
the only blocker for the creation of v0.1 of the merged dataset is #531, which is pretty close to be merged. At that point we will be able to do the following:
1) create a dataset without a table writer (essentially a view) with a columnSet that resembles the external columns. This dataset, does not have storage on its own so it is trivial from the ops point of view.
2) add a simple abstraction that extends RelationalSource and switches between tables (this is optional)
3) add QueryProcessor that resolves external columns (those in the columnset) to a table and internal columns
4) fix all the corner cases that will come up
5) Start using it and cleanup the dataset switch from sentry.

The abstractions around the query in Snuba are a little rough as of now, but we already have some basic ones and they are going to improve soon, so I think it is better to go this way instead of building yet another Query abstraction for this topic. (to be fair I think a Query abstraction in sentry would be a good idea for other reasons, but it should not be a driver for this)

**Comment by markstory** [comment]
> Is there a reason we can't resolve the dataset based on the internal names now? 

The main blocker on resolving the dataset based on internal names is that we can't get to the internal names without knowing the dataset, which we have to infer from the public names used. For example the fields `transaction.duration, transaction.op` could both be tags on events, or real columns on the transactions dataset. Until we decide the dataset we can't build the internal names.

@fpacifici I see a Query abstraction in sentry helping reduce the dictionary spreads and mutations we do in multiple places. There is the beginnings of such a wrapper in `utils.snuba.SnubaQueryParams` but it isn't broadly used yet.

Having a merged dataset that is more closely aligned to the external interface will remove a considerable amount of complexity from sentry albeit by moving it snuba. But I think that is a better place for it to be as it is closer to the storage and processing logic.

File: src/sentry/utils/snuba.py (hunk 1 above)
Comment (lynnagara) [reply src/sentry/utils/snuba.py]: Do we need to fix `get_function_index` then? IIRC "IN" condition shouldn't quack like a `get_function_index` since function expression should always have a list/tuple in the second position and where IN just has the string "IN"

Comment (dashed) [reply src/sentry/utils/snuba.py]: I think the answer is yes. 

I think @markstory mentioned to me that this may be resolved in this PR, or in an independent PR.

Comment (markstory) [reply src/sentry/utils/snuba.py]: I can do a follow up pull request to fix `IN` being treated as a function.

**Comment by fpacifici** [comment]
> I see a Query abstraction in sentry helping reduce the dictionary spreads and mutations we do in multiple places. 

I like the idea.

File: src/sentry/eventstore/snuba/backend.py [hunk 2]
```diff
@@ -32,10 +32,9 @@ def get_events(
         Get events from Snuba.
         """
         assert filter, "You must provide a filter"
-
         cols = self.__get_columns(additional_columns)
 
-        result = snuba.raw_query(
+        result = snuba.dataset_query(
```
Comment (lynnagara) [inline src/sentry/eventstore/snuba/backend.py]: There is a risk of making the wrong dataset decision here if a user has got a tag that is a transaction column

Comment (markstory) [reply src/sentry/eventstore/snuba/backend.py]: Good point. I'll coerce the dataset here.

Comment (lynnagara) [reply src/sentry/eventstore/snuba/backend.py]: I think we also need to do this everywhere we call `eventstore.get_events`. Similar to https://github.com/getsentry/sentry/pull/15254

Comment (markstory) [reply src/sentry/eventstore/snuba/backend.py]: We can't do it everywhere as there are a handful of endpoints that need to access both types of events.

We could add `dataset` onto `eventstore.Filter` as you/manu had in another pull request. This would let us force the correct dataset in other error only endpoints.

Comment (lynnagara) [reply src/sentry/eventstore/snuba/backend.py]: We can get around this by just including the condition `["type", "!=", "transaction"]` anywhere this gets called where we want to force the right dataset.

File: src/sentry/eventstore/snuba/backend.py [hunk 3]
```diff
@@ -160,9 +159,7 @@ def __get_event_id_from_filter(self, filter=None, orderby=None):
             limit=1,
             referrer="eventstore.get_next_or_prev_event_id",
             orderby=orderby,
-            dataset=snuba.detect_dataset(
-                {"conditions": filter.conditions}, aliased_conditions=True
-            ),
+            dataset=snuba.detect_dataset({"conditions": filter.conditions}),
```
Comment (lynnagara) [inline src/sentry/eventstore/snuba/backend.py]: Note that this changes the dataset detection logic to not take into account any selected columns. Intentional?

Comment (markstory) [reply src/sentry/eventstore/snuba/backend.py]: The `event_id` and `project_id` columns are present on all datasets.

Comment (lynnagara) [reply src/sentry/eventstore/snuba/backend.py]: Sorry, I misread the diff and thought the was in get_events

File: src/sentry/api/endpoints/organization_eventid.py [hunk 4]
```diff
@@ -46,12 +46,11 @@ def get(self, request, organization, event_id):
...
-            event = eventstore.get_events(
-                filter=eventstore.Filter(
-                    project_ids=project_slugs_by_id.keys(), event_ids=[event_id]
-                ),
-                limit=1,
-            )[0]
+            snuba_filter = eventstore.Filter(
+                project_ids=project_slugs_by_id.keys(), event_ids=[event_id]
+            )
+            snuba_filter.conditions = [["event.type", "!=", "transaction"]]
```
Comment (lynnagara) [inline src/sentry/api/endpoints/organization_eventid.py]: nit: You can pass conditions on the line above where Filter is instantiated

**Review by lynnagara — APPROVED** [review]
LGTM
